packages/i18n-calypso/types/index.d.ts

```diff
@@ -0,0 +1,87 @@
+// Type definitions for i18n-calypso
+// Project: i18n-calypso
+
+/**
+ * External dependencies
+ */
+import { ComponentClass, ComponentType, ReactElement } from 'react';
+import moment from 'moment';
+
+declare namespace i18nCalypso {
+	export type Substitutions = string | string[] | { [placeholder: string]: string };
+
+	export interface ComponentInterpolations {
+		[placeholder: string]: ReactElement;
+	}
+
+	export interface TranslateOptions {
+		/**
+		 * Arguments you would pass into sprintf to be run against the text for string substitution.
+		 */
+		args?: Substitutions;
+
+		/**
+		 * Components for interpolation in the provided string.
+		 */
+		components?: ComponentInterpolations;
+
+		/**
+		 * Comment that will be shown to the translator for anything that may need to be explained about the translation.
+		 */
+		comment?: string;
+
+		/**
+		 * Provides the ability for the translator to provide a different translation for the same text in two locations (dependent on context). Usually context should only be used after a string has been discovered to require different translations. If you want to provide help on how to translate (which is highly appreciated!), please use a comment.
+		 */
+		context?: string;
+	}
+
+	export type TranslatedString = string;
+
+	export function translate( oringinal: string ): TranslatedString;
+	export function translate(
+		oringinal: string,
+		plural: string,
+		options: TranslateOptions & { count: number }
+	): TranslatedString;
+	export function translate( oringinal: string, options: TranslateOptions ): TranslatedString;
+	export function translate(
+		oringinal: string,
+		plural: string,
+		options: TranslateOptions
+	): TranslatedString;
+
+	export function hasTranslation( original: string ): boolean;
+
+	export interface NumberFormatOptions {
+		decimals?: number;
+		decPoint?: string;
+		thousandsSep?: string;
+	}
+
+	export function numberFormat( number: number, numberOfDecimalPlaces: number );
+	export function numberFormat( number: number, options: NumberFormatOptions );
+
+	export interface LocalizeProps {
+		translate: typeof translate;
+		numberFormat: typeof numberFormat;
+		moment: typeof moment;
+	}
+
+	// Infers prop type from component C
+	export type GetProps< C > = C extends ComponentType< infer P > ? P : never;
+
+	export type WithoutLocalizedProps< OrigProps > = Pick<
+		OrigProps,
+		Exclude< keyof OrigProps, keyof LocalizeProps >
+	>;
+
+	export type LocalizedComponent< C > = ComponentClass< WithoutLocalizedProps< GetProps< C > > >;
+
+	export function localize< C >( component: C ): LocalizedComponent< C >;
```

Comment:
This was a bit tricky, I took _some_ inspiration from [`react-redux`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/e253d335c1f036e67ace2898b598a3f620a6bb78/types/react-redux/index.d.ts#L107-L113).

Seems to work well!

```tsx
import { localize, useTranslate } from 'i18n-calypso';
class C extends React.Component< { translate: ( s: string ) => string; n: 1 } > {
	render() {
		const t = useTranslate();
		return <p>{ t( 'Hello ' ) + this.props.translate( 'world!' ) }</p>;
	}
}

<C />; // Type '{}' is missing the following properties from type 'Readonly<{ translate: (s: string) => string; n: 1; }>': translate, n

const LC = localize( C );
<LC n={ 2 } />; // Type '2' is not assignable to type '1'.
<LC n={ 1 } />; // 👍 😎
```

Reply:
This is a bit over me :see_no_evil:  but it seems to be working well :sweat_smile: